File: client/smart_lock/lib/models/task.dart
class Task {
  final String id;
  final String title;
  final String description;
  final String assignedTo;
  final String assignedBy;
  final String status;
  final String priority;
  final DateTime dueDate;
  final DateTime? completedAt;
  final String? notes;
  final List<Map<String, String>>? attachments;
  final List<String>? relatedRequests;

  Task({
    required this.id,
    required this.title,
    required this.description,
    required this.assignedTo,
    required this.assignedBy,
    required this.status,
    required this.priority,
    required this.dueDate,
    this.completedAt,
    this.notes,
    this.attachments,
    this.relatedRequests,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['_id'],
      title: json['title'],
      description: json['description'],
      assignedTo: json['assignedTo'],
      assignedBy: json['assignedBy'],
      status: json['status'],
      priority: json['priority'],
      dueDate: DateTime.parse(json['dueDate']),
      completedAt: json['completedAt'] != null
          ? DateTime.parse(json['completedAt'])
          : null,
      notes: json['notes'],
      attachments: json['attachments'] != null
          ? List<Map<String, String>>.from(json['attachments'])
          : null,
      relatedRequests: json['relatedRequests'] != null
          ? List<String>.from(json['relatedRequests'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'assignedTo': assignedTo,
      'status': status,
      'priority': priority,
      'dueDate': dueDate.toIso8601String(),
      'notes': notes,
      'attachments': attachments,
      'relatedRequests': relatedRequests,
    };
  }
}

File: server/src/models/Task.js
const mongoose = require('mongoose');

const taskSchema = new mongoose.Schema({
  workerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Task must be assigned to a worker'],
  },
  title: {
    type: String,
    required: [true, 'Please provide a task title'],
    trim: true,
  },
  description: {
    type: String,
    required: [true, 'Please provide a task description'],
  },
  status: {
    type: String,
    enum: ['Pending', 'In Progress', 'Completed', 'Cancelled'],
    default: 'Pending',
  },
  priority: {
    type: String,
    enum: ['Low', 'Medium', 'High', 'Urgent'],
    default: 'Medium',
  },
  dueDate: {
    type: Date,
    required: [true, 'Please specify the due date'],
  },
  assignedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'Task must have an assigner'],
  },
  relatedRequests: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'AccessLog',
  }],
  notes: {
    type: String,
    trim: true,
  },
  completedAt: {
    type: Date,
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true },
});

// Indexes for better query performance
taskSchema.index({ workerId: 1, status: 1 });
taskSchema.index({ dueDate: 1 });
taskSchema.index({ priority: 1 });

// Virtual populate with worker details
taskSchema.virtual('worker', {
  ref: 'User',
  foreignField: '_id',
  localField: 'workerId',
  justOne: true,
});

// Virtual populate with assigner details
taskSchema.virtual('assigner', {
  ref: 'User',
  foreignField: '_id',
  localField: 'assignedBy',
  justOne: true,
});

const Task = mongoose.model('Task', taskSchema);

module.exports = { Task }; 